lib/hls/hls_parser.js

```diff
@@ -2305,6 +2311,46 @@ shaka.hls.HlsParser = class {
           references.push(reference);
         }
       }
+
+      const segmentDateRangeTags = shaka.hls.Utils.filterTagsByName(
+          item.tags, 'EXT-X-DATERANGE');
+      const firstProgramDateTimeTag = shaka.hls.Utils.getFirstTagWithName(
+          playlist.segments[0].tags, 'EXT-X-PROGRAM-DATE-TIME');
+      const offsetTime = this.calculateOffsetTime_(firstStartTime,
+          firstProgramDateTimeTag);
+
+      for (const dateRangeTag of segmentDateRangeTags) {
+        const newRegion = this.parseDateRangeTag_(dateRangeTag, offsetTime);
+
+        if (newRegion.id && this.regionIdToRegionInfosMap_.has(newRegion.id)) {
+          const existingRegion = this.regionIdToRegionInfosMap_
+              .get(newRegion.id);
+
+          if (existingRegion.endTime !== newRegion.endTime) {
+            existingRegion.endTime = newRegion.endTime;
+
+            const existingDateRange = existingRegion.hlsDateRangeInfo;
+            const newDateRange = newRegion.hlsDateRangeInfo;
+
+            if (existingDateRange && newDateRange) {
+              // The DURATION tag should be final, if it was not set before
+              // and it is set now, update the region's duration
+              if (newDateRange.duration) {
+                existingDateRange.duration = newDateRange.duration;
+              } else if (newDateRange.plannedDuration) {
+                existingDateRange.plannedDuration = newDateRange
+                    .plannedDuration;
+              }
+            }
+
+            // Update this region because the endTime (duration) has changed
+            this.playerInterface_.onTimelineRegionAdded(existingRegion);
+          }
+        } else {
+          this.playerInterface_.onTimelineRegionAdded(newRegion);
+          this.regionIdToRegionInfosMap_.set(newRegion.id, newRegion);
```

Comment:
Do we need to remove old entries from regionIdToRegionInfosMap_ after they moved out of DVR window?

Reply:
Yes, they should be removed. Added the checking for removal.